File: fsharp/transpose/Transpose.fs
module Transpose

open System

let rec transposeChars = function
    | (_::_)::_ as M -> List.map List.head M :: transposeChars (List.map List.tail M)
    | _ -> []

let transpose (input: string list) = 
    if List.length input = 0 then []
    else
        let maxlen = input |> Seq.map (fun word -> word.Length) |> Seq.max
        let charMatrix = input |> List.map (fun word ->  word.PadRight(maxlen, '*')) |> List.map (fun word -> word.ToCharArray() |> Array.toList)
        transposeChars charMatrix 
        |> Seq.map (fun charList -> charList |> Array.ofList |> String) 
        |> Seq.map (fun word -> word.TrimEnd('*').Replace('*', ' '))
        |> Seq.toList

    

File: csharp/transpose/Transpose.cs
using System.Linq;
using System.Collections.Generic;

public static class Transpose
{
    public static string String(string input)
    {
        if (input == "") return "";
        var len = input.Split('\n').Select(word => word.Length).Max();
        var charMatrix = input.Split('\n').Select(word => word.PadRight(len, '*')).ToArray();
        var transpose = new List<System.String>();
        for (var i = 0; i < len; i++)
        {
            var line = new System.String("");
            for (var j = 0; j < charMatrix.Length; j++)
            {
                line += charMatrix[j][i];
            }
            transpose.Add(line.TrimEnd('*').Replace('*', ' '));
        }
        return System.String.Join("\n", transpose);
    }
}